exec/native_compile_test.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2019 The go-interpreter Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build !appengine
+
+package exec
+
+import (
+	"bytes"
+	"runtime"
+	"testing"
+
+	"github.com/go-interpreter/wagon/disasm"
+	"github.com/go-interpreter/wagon/exec/internal/compile"
+	ops "github.com/go-interpreter/wagon/wasm/operators"
+)
+
+func fakeNativeCompiler(t *testing.T) *nativeCompiler {
+	t.Helper()
+	return &nativeCompiler{
+		Builder:   &mockInstructionBuilder{},
+		Scanner:   &mockSequenceScanner{},
+		allocator: &mockPageAllocator{},
+	}
+}
+
+type mockSequenceScanner struct {
+	emit []compile.CompilationCandidate
+}
+
+func (s *mockSequenceScanner) ScanFunc(bc []byte, meta *compile.BytecodeMetadata) ([]compile.CompilationCandidate, error) {
+	return s.emit, nil
+}
+
+type mockPageAllocator struct{}
+
+func (a *mockPageAllocator) AllocateExec(asm []byte) (compile.NativeCodeUnit, error) {
+	return nil, nil
+}
+
+func (a *mockPageAllocator) Close() error {
+	return nil
+}
+
+type mockInstructionBuilder struct{}
+
+func (b *mockInstructionBuilder) Build(candidate compile.CompilationCandidate, code []byte, meta *compile.BytecodeMetadata) ([]byte, error) {
+	return []byte{byte(candidate.Start), byte(candidate.End)}, nil
+}
+
+func TestNativeAsmStructureSetup(t *testing.T) {
+	nc := fakeNativeCompiler(t)
+
+	constInst, _ := ops.New(ops.I32Const)
+	addInst, _ := ops.New(ops.I32Add)
+	subInst, _ := ops.New(ops.I32Sub)
+	setGlobalInst, _ := ops.New(ops.SetGlobal)
+
+	wasm, err := disasm.Assemble([]disasm.Instr{
+		{Op: constInst, Immediates: []interface{}{int32(1)}},
+		{Op: constInst, Immediates: []interface{}{int32(1)}},
+		{Op: addInst},
+		{Op: setGlobalInst, Immediates: []interface{}{uint32(0)}},
+
+		{Op: constInst, Immediates: []interface{}{int32(8)}},
+		{Op: constInst, Immediates: []interface{}{int32(16)}},
+		{Op: constInst, Immediates: []interface{}{int32(4)}},
+		{Op: addInst},
+		{Op: subInst},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	vm := &VM{
+		funcs: []function{
+			compiledFunction{
+				code: wasm,
+			},
+		},
+		nativeBackend: nc,
+	}
+	vm.newFuncTable()
+
+	// setup mocks.
+	nc.Scanner.(*mockSequenceScanner).emit = []compile.CompilationCandidate{
+		// Sequence with single integer op - should not compiled.
+		compile.CompilationCandidate{Start: 0, End: 7, EndInstruction: 3, Metrics: compile.Metrics{IntegerOps: 1}},
```

Comment:
looking at the history of the changes, it doesn't seem to me `EndInstruction`'s value was changed here, nor below.
shouldn't they be (resp.) `4` and `10` ?

Reply:
Fixed.

Note: Start/EndInstruction is not part of the code under test, this test just tests `tryNativeCompile()`, and mocks everything else.